src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/entity/Licence.kt

```diff
@@ -187,4 +188,39 @@ data class Licence(
     }
     return false
   }
+
+  fun earliestReleaseDate(listOfBankHolidays: List<LocalDate>, workingDays: Int): LocalDate? {
+    val releaseDate = actualReleaseDate ?: conditionalReleaseDate
+    val possibleReleaseDates: MutableList<LocalDate> = mutableListOf()
+
+    if (releaseDate !== null && isBankHolidayOrWeekend(listOfBankHolidays, releaseDate)) {
+      return getEarliestReleaseDate(listOfBankHolidays, workingDays, releaseDate, possibleReleaseDates)
+    }
+    return releaseDate
+  }
+
+  /** Friday is also considered as weekend */
+  private fun isBankHolidayOrWeekend(listOfBankHolidays: List<LocalDate>, releaseDate: LocalDate?): Boolean {
+    val dayOfWeek = releaseDate?.dayOfWeek
+    if (dayOfWeek == DayOfWeek.FRIDAY || dayOfWeek == DayOfWeek.SATURDAY || dayOfWeek == DayOfWeek.SUNDAY) {
+      return true
+    }
+    return listOfBankHolidays.contains(releaseDate)
+  }
+
+  private fun getEarliestReleaseDate(
+    listOfBankHolidays: List<LocalDate>,
+    workingDays: Int,
+    normalReleaseDate: LocalDate?,
+    possibleReleaseDates: MutableList<LocalDate>,
+  ): LocalDate {
+    val earlyPossibleDate = normalReleaseDate?.minusDays(1)
+    if (possibleReleaseDates.size < workingDays) {
+      if (earlyPossibleDate !== null && !isBankHolidayOrWeekend(listOfBankHolidays, earlyPossibleDate)) {
+        possibleReleaseDates.add(earlyPossibleDate)
+      }
+      getEarliestReleaseDate(listOfBankHolidays, workingDays, earlyPossibleDate, possibleReleaseDates)
+    }
+    return possibleReleaseDates[workingDays - 1]
```

Comment:
I'm not sure your type sig is correct - how can you return a non nullable date if your release date is nullable?

rather than using recursion, you could use [kotlin magic](https://kotlinlang.org/docs/sequences.html#from-a-function) to model it as a stream of dates..

something like:
```
val releaseDates = generateSequence(releaseDate) { it.minusDays(1) }
    .filterNot { isBankHoliday(it) }
    .take(workingDays)
    .reverse()
    .firstOrNull()
```

Not sure that logic is quite right... but hopefully close to what you want?

Reply:
If release date is null I am returning it immediately. If not only then we are calling getEarliestReleaseDate method.